### PR TITLE
Build updates and modernization

### LINKED ISSUE #370: Tests skipped with "skipped 'Test requires a non-null GUI backend'" even with toolkit installed
Running the test suite, even with a gui toolkit installed, results in 5 tests being skipped with the message "skipped 'Test requires a non-null GUI backend'".  It seems you have to manually set the `ETS_TOOLKIT` variable. 
```
test_gui_trait_expects_IGUI_interface (ui.tasks.tests.test_tasks_application.TestTasksApplication) ... skipped 'Test requires a non-null GUI backend'
test_layout_load (ui.tasks.tests.test_tasks_application.TestTasksApplication) ... skipped 'Test requires a non-null GUI backend'
test_layout_load_pickle_protocol_3 (ui.tasks.tests.test_tasks_application.TestTasksApplication) ... skipped 'Test requires a non-null GUI backend'
test_layout_save_uses_protocol_2 (ui.tasks.tests.test_tasks_application.TestTasksApplication) ... skipped 'Test requires a non-null GUI backend'
test_layout_save_with_protocol_3 (ui.tasks.tests.test_tasks_application.TestTasksApplication) ... skipped 'Test requires a non-null GUI backend'
```

environment details
<details>

```
$ edm list
# Packages in environment 'test-envisage3'
#  prefix: '/Users/aayres/.edm/envs/test-envisage3'
appdirs            1.4.3-1   enthought/free
distribute_remove  1.0.0-4   enthought/free
icu4c              50.2-2    enthought/free
libjpeg            7.0-5     enthought/free
libpng             1.6.36-5  enthought/free
libxml2            2.9.7-1   enthought/free
libxslt            1.1.30-1  enthought/free
packaging          16.8-4    enthought/free
pip                20.0.2-2  enthought/free
pyparsing          2.2.0-1   enthought/free
pyqt5              5.14.2-3  enthought/free
pyqt5_sip          12.7.2-1  enthought/free
qt                 5.12.6-6  enthought/free
setuptools         41.6.0-4  enthought/free
six                1.15.0-1  enthought/free
```

```
$ pip list
Package            Version 
------------------ --------
appdirs            1.4.3   
appnope            0.1.0   
apptools           4.5.0   
backcall           0.2.0   
configobj          5.0.6   
coverage           5.3     
decorator          4.4.2   
distribute-remove  1.0.0   
envisage           5.0.0rc1
flake8             3.8.4   
icu4c              50.2    
importlib-metadata 3.1.0   
ipykernel          5.3.4   
ipython            7.16.1  
ipython-genutils   0.2.0   
jedi               0.17.2  
jupyter-client     6.1.7   
jupyter-core       4.7.0   
libjpeg            7.0     
libpng             1.6.36  
libxml2            2.9.7   
libxslt            1.1.30  
mccabe             0.6.1   
packaging          16.8    
parso              0.7.1   
pexpect            4.8.0   
pickleshare        0.7.5   
pip                20.0.2  
prompt-toolkit     3.0.8   
ptyprocess         0.6.0   
pycodestyle        2.6.0   
pyface             7.1.0   
pyflakes           2.2.0   
Pygments           2.7.2   
pyparsing          2.2.0   
PyQt5              5.14.2  
PyQt5-sip          12.7.2  
python-dateutil    2.8.1   
pyzmq              20.0.0  
Qt                 5.12.6  
setuptools         41.6.0  
six                1.15.0  
tornado            6.1     
traitlets          4.3.3   
traits             6.1.1   
traitsui           7.1.0   
wcwidth            0.2.5   
zipp               3.4.0
```

</details>

### LINKED ISSUE #511: Add Python 3.11 compatibility
The upcoming 7.0 release would be a good time to make sure that the Envisage codebase has no compatibility issues with Python 3.11.

Currently known blockers:

- [ ] When using PySide6, we currently depend on PySide6 < 6.4.0, and that's not available for Python 3.11. There's ongoing work in Pyface to remove this restriction.

### LINKED ISSUE #512: Drop Python 3.6 support
Pyface is dropping Python 3.6 support, so we'll need to too. This should happen before the upcoming release.

### LINKED ISSUE #519: Use pyproject.toml
This is a pure Python project with no extensions or specialist build needs, so it should be straightforward to convert from using `setup.py` and `setup.cfg` to `pyproject.toml` and `.flake8`.

### LINKED ISSUE #525: Drop wxPython support in `etstool.py`
The `wxPython` support in `etstool.py` involves installing wxPython from external sources, and as such is rather fragile. For example:

- on macOS, a `pip install wxPython` fails because it tries to build wxPython from source, and that requires `attrdict3` (see https://github.com/wxWidgets/Phoenix/issues/2225)
- after installing `attrdict3`, the wxPython installation from source takes a very long time; I gave up after approximately 40 minutes
- `wxPython` does _have_ wheels, but on macOS those are incompatible with EDM's Python (I think because the wheels don't support versions of macOS older than 11.0)
- on Linux, we're forced to install from `extras.wxPython.org`

Envisage doesn't contain much toolkit-specific code, and we can test with `wxPython` on non-EDM installs, so there seems little value keeping the wxPython support in `etstool.py`. Let's keep `etstool.py` for testing with the toolkits that _are_ available in EDM (currently PySide6 and PyQt6).

### PR DESCRIPTION
**Pull request by mdickinson**
This PR updates the build machinery to use pyproject.toml, and drops Python 3.6 support.

## Detailed changes

- Drop Python 3.6 support; we now require Python >= 3.7.
- Drop Python 3.6 from test workflows.
- Only use PySide6 for testing. (The main package still permits PySide2 and PyQt5 for Python versions for which they're available; we just don't test on those.)
- Rename `setup.cfg` (which only contained `flake8` configuration settings) to `.flake8`
- Replace `setup.py` with `pyproject.toml`
- Remove various outdated references to `setup.py`
- Remove `envisage.version` and `envisage.__version__` (and tests)
- Optional dependencies (a.k.a. `extras_require`) have been removed
- We no longer rely on `ETS_TOOLKIT` (or `QT_API`) being set in order to run tests that require a GUI; we just rely on Pyface's default mechanisms for toolkit detection. (As a consequence, we retain the option to _use_ `ETS_TOOLKIT` if it's helpful.)
- `etstool.py` support for backends other than those available through EDM has been removed. Meanwhile, support for `pyqt6` has been added. (Tests currently fail with PyQt6, due to a known issue in TraitsUI: https://github.com/enthought/traitsui/issues/1926)

Closes #370
Closes #511
Closes #512
Closes #519
Closes #525

### CODE REVIEW

**Comment by mdickinson** [comment]
@corranwebster This is ready for review, if you have the bandwidth available.

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,95 @@
...
+]
+
+[project.entry-points.'envisage.plugins']
+'envisage.core' = 'envisage.core_plugin:CorePlugin'
+
+[project.entry-points.etsdemo_data]
+demo_examples = 'envisage.examples._etsdemo_info:info'
+
+[project.optional-dependencies]
+pyside6 = ['pyside6<6.4.0']
```
Comment (mdickinson) [inline pyproject.toml:46]: I'm not totally convinced that this is useful. It may make more sense to remove it.

Comment (mdickinson) [reply pyproject.toml:46]: Thinking a bit more: this _is_ useful in that it records a necessary version restriction on PySide6, and that restriction isn't recorded anywhere else. However, that version restriction really belongs in Pyface, not here - it's Pyface that's the source of the incompatibility with PySide 6.4. Moreover, the next release of Pyface will be removing that restriction.

I think we should assume that we won't release Envisage 7.0 until after the next release of Pyface. In that case, we can remove this section of the `pyproject.toml` right now (and in the meantime, while we're waiting for the Pyface release, we can modify the GitHub Actions workflow to work with a supported version of PySide6).

**Comment by mdickinson** [comment]
> This is ready for review, if you have the bandwidth available.

Sorry, I take that back. The juggling with the changes in #526 is getting annoying; I think it makes most sense to merge that PR into this one. I'll change back to draft while that's in progress.

File: pyproject.toml (hunk 1 above)
Comment (mdickinson) [reply pyproject.toml]: Done in 0ea87cfa193cffa42b89eb04c2e3180e833a708a

**Review by corranwebster — APPROVED** [review]

**Pull request merged**